### PR TITLE
refactor `sq.gr.moran`, remove pysal depend

### PR DESCRIPTION
**Pull request by giovp**
### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (change that would cause existing functionality to not work as before)
- [x] New feature (non-breaking change which adds functionality)

### Description
Re-implement Moran's I to remove libpysal dependency

### How has this been tested?
not yet

### Closes

TODO:

- [ ] add row wise transform?
- [ ] improve speed ?
- [ ] change docs and reference

@michalk8 implementation is really trivial see [wiki](https://en.wikipedia.org/wiki/Moran%27s_I) and original code https://github.com/pysal/esda/blob/642880e0b49913de07e5b7fafb3dd0b5855ed2ef/esda/moran.py#L234

unfortunately, it's slower than original, not sure why. one interpretation could be that numba vectorized dot product is still slower than `np.dot` and the gain in permutation is lost. do you see any major performance issues? otherwise could also revert to plain numpy.
anyway, this will allow us to remove pysal+esda from dependency. top 100 genes by pysal are consistent
image

### CODE REVIEW

File: squidpy/gr/_ppatterns.py [hunk 1]
```diff
@@ -251,20 +234,54 @@ def _moran_helper(
...
+    z2ss = (z * z).sum()
+    return len(counts) / data.sum() * inum / z2ss
 
 
-def _set_weight_class(adata: AnnData, key: str) -> W:
-    X = adata.obsp[key]
-    if not isspmatrix_lil(X):
-        X = X.tolil()
+@njit(
+    ft[:](ft[:], it[:], it[:], ft[:], it),
```
Comment (ivirshup) [inline squidpy/gr/_ppatterns.py:263]: Indices of sparse matrices are not always `int32`, if you have enough entries they will become `int64`.

File: squidpy/gr/_ppatterns.py [hunk 2]
```diff
@@ -251,20 +234,54 @@ def _moran_helper(
     return pd.DataFrame(moran_list, columns=["I", "pval_sim", "VI_sim"], index=gen)
 
 
-def _compute_moran(y: np.ndarray, w: W, transformation: str, permutations: int) -> Tuple[float, float, float]:
-    mi = esda.moran.Moran(y, w, transformation=transformation, permutations=permutations)
-    return mi.I, mi.p_z_sim, mi.VI_sim
+@njit(
+    ft(ft[:], it[:], it[:], ft[:]),
+    parallel=False,
```
Comment (ivirshup) [inline squidpy/gr/_ppatterns.py:239]: In theory, is parallelizable, but I've run into some issues with this recently

File: squidpy/gr/_ppatterns.py (hunk 1 above)
Comment (ivirshup) [reply squidpy/gr/_ppatterns.py:263]: Also, probably good to collect to `float64` if you're expecting a lot of samples.

File: squidpy/gr/_ppatterns.py [hunk 3]
```diff
@@ -239,7 +218,11 @@ def _moran_helper(
 
     moran_list = []
     for g in gen:
-        mi = _compute_moran(adata.obs_vector(g, layer=layer), weights, transformation, permutations)
+        counts = adata.obs_vector(g, layer=layer).copy().astype(fp)
```
Comment (ivirshup) [inline squidpy/gr/_ppatterns.py:221]:
```suggestion
        counts = adata.obs_vector(g, layer=layer).astype(fp, copy=True)
```

**Review by ivirshup — COMMENTED** [review]
Hey, saw this and had a couple points.

I wouldn't be surprised if the overhead of joblib was significant here. I could give a few more pointers on performance if you want (I'd have to look closer, but I think there's a few computations that are repeated more than needed).

Any relationship to https://github.com/theislab/scanpy/issues/1698?

**Comment by giovp** [comment]
thanks a lot @ivirshup ! so re performance, I calculated including joblib in both cases (their original implementation and mine with numba). It's slower nonetheless. I was wondering if there is some improvement I could do on the numba code itself, I think using joblib is necessary anyway to parallelize across genes.

re theislab/scanpy#1698 yeah thanks for pointing it out, so one option is to submit this PR directly there, and wrap `scanpy.metrics` here. Could also wrap your implementation of geary's C there if you agree. 

would like to get feedback on this from @michalk8 @hspitzer : shall we move this to scanpy metrics module and wrap it as moran here?

File: squidpy/gr/_ppatterns.py [hunk 4]
```diff
@@ -251,20 +234,54 @@ def _moran_helper(
...
+    zl = np.empty(len(z))
+    N = len(indptr) - 1
+    for i in range(N):
+        s = slice(indptr[i], indptr[i + 1])
+        i_indices = indices[s]
+        i_data = data[s]
+        zl[i] = np.sum(i_data * z[i_indices])
+    inum = (z * zl).sum()
+    z2ss = (z * z).sum()
+    return len(counts) / data.sum() * inum / z2ss
```
Comment (ivirshup) [inline squidpy/gr/_ppatterns.py:259]: `counts - counts.mean()`, `z2ss = (z * z).sum()`, and `data.sum()` should be the same w.r.t. each permutation, so you can probably compute these outside the permutation loop and pass them in.

**Review by ivirshup — COMMENTED** [review]
I've commented a bit more on the performance, basically some things that you're computing each permutation that I think should be constant across permutations.

Could you share the cases you're using to benchmark?

**Comment by ivirshup** [comment]
> re scanpy#1698 yeah thanks for pointing it out, so one option is to submit this PR directly there, and wrap scanpy.metrics here. Could also wrap your implementation of geary's C there if you agree.

Whatever works for you. I think it could also make sense to have the code live both places for a bit depending on how you feel about requiring newest scanpy/ release schedules.

File: squidpy/gr/_ppatterns.py (hunk 1 above)
Comment (giovp) [reply squidpy/gr/_ppatterns.py:263]: done, thank you!

File: squidpy/gr/_ppatterns.py (hunk 2 above)
Comment (giovp) [reply squidpy/gr/_ppatterns.py:239]: ok interesting, but didn't see much improvement with parallel anyway

File: squidpy/gr/_ppatterns.py (hunk 3 above)
Comment (giovp) [reply squidpy/gr/_ppatterns.py:221]: done, thank you!

File: squidpy/gr/_ppatterns.py (hunk 4 above)
Comment (giovp) [reply squidpy/gr/_ppatterns.py:259]: good point, done but still not much better

**Comment by giovp** [comment]
```python
adata = sq.datasets.visium_hne_adata()
sq.gr.spatial_neighbors(adata)
genes = adata[:, adata.var.highly_variable].var_names.values[0:100]

# numba
sq.gr.moran(adata, genes=genes, n_jobs=1)
# time 01:30
moran_numba = adata.uns["moranI"].copy()

# switch to master, pysal
sq.gr.moran(adata, genes=genes, n_jobs=1)
# time 00:18
moran_pysal = adata.uns["moranI"].copy()

g = sns.scatterplot(moran_pysal["I"], moran_numba["I"])
g.set_xlabel("pysal")
g.set_ylabel("numba")
```
image

re moving to scanpy, brief talk with @hspitzer and she is fine with moving it there, I think with the ratio of scanpy being core of squidpy it makes sense. 
We would than wrap both this and geary's c in `sq.gr.moran` and `sq.gr.gearysc` for convenience to handle spatial neighbors. How does it sounds @ivirshup @michalk8 ?

File: squidpy/gr/_ppatterns.py [hunk 5]
```diff
@@ -239,7 +218,11 @@ def _moran_helper(
 
     moran_list = []
     for g in gen:
-        mi = _compute_moran(adata.obs_vector(g, layer=layer), weights, transformation, permutations)
+        counts = adata.obs_vector(g, layer=layer).astype(fp, copy=True)
```
Comment (michalk8) [inline squidpy/gr/_ppatterns.py:221]: This can be moved out of the loop.

File: squidpy/gr/_ppatterns.py [hunk 6]
```diff
@@ -251,20 +234,61 @@ def _moran_helper(
...
 
+    perms = np.empty(n_perms, dtype=ft)
+    res = np.empty(3, dtype=ft)
 
-def _set_weight_class(adata: AnnData, key: str) -> W:
-    X = adata.obsp[key]
-    if not isspmatrix_lil(X):
-        X = X.tolil()
+    z = counts - counts.mean()
+    data_sum = data.sum()
```
Comment (michalk8) [inline squidpy/gr/_ppatterns.py:281]: `data_sum` can be computed in the caller.

**Comment by michalk8** [comment]
>  We would than wrap both this and geary's c in sq.gr.moran and sq.gr.gearysc for convenience to handle spatial neighbors. How does it sounds @ivirshup @michalk8 ?

Sounds good to me. In terms of performance, 2 things come to mind:
- look at the esda implementation
- switch the joblib par. loops (perms/genes instead of genes/perms)
- maybe try implementing the crucial parts in Cython? not sure if there would be a speed-up

**Comment by giovp** [comment]
this was addressed in #317

**Pull request closed without merging**